File: modulefederation/libs/portal/src/lib/tracer/tracer.ts
import { TraceLevel } from './trace-level.enum';
import { Trace } from './trace';
import { TraceEntry } from './trace-entry';
import { Injectable } from "@angular/core";
import { TracerModule } from './tracer.module';
import { StackFrame, Stacktrace } from '../common/util';


/**
 * A Tracer is used to emit trace messages for development purposes.
 * While it shares the logic of a typical logger, it will be turned of in production.
 */
@Injectable({
    providedIn: 'root'
})
export class Tracer {
    // static

    public static ENABLED = true

    public static instance : Tracer

    // instance data

    private traceLevels : { [path : string] : TraceLevel } = {}
    private cachedTraceLevels : { [path : string] : TraceLevel } = {}
    private modifications = 0
    private sink : Trace | undefined

    // constructor

    constructor() {
        const tracerConfiguration = TracerModule.tracerConfiguration

        // enabled

        Tracer.ENABLED = tracerConfiguration.enabled

        // some more

        this.sink = tracerConfiguration.trace

        // set paths

        if (tracerConfiguration.paths)
            for (const path of Object.keys(tracerConfiguration.paths))
                this.setTraceLevel(path, tracerConfiguration.paths[path])

        Tracer.instance = this
    }

    static getInstance() {
        if (!Tracer.instance)
            new Tracer()

        return Tracer.instance
    }

    // public

    public static Trace(path : string, level : TraceLevel, message : string, ...args : any[]) {
        const instance = Tracer.getInstance()

        if ( instance.getTraceLevel(path) >= level) {
            const stack = new Error().stack!

            const frames = Stacktrace.createFrames(stack)

            const lastFrame = frames[1]

            instance.trace(path, level, message, lastFrame, ...args)
        }
    }

    // public

    public isTraced(path : string, level : TraceLevel) : boolean {
        return Tracer.ENABLED && this.getTraceLevel(path) >= level
    }

    public async trace(path : string, level : TraceLevel, message : string, frame: StackFrame,  ...args : any[]) {
        if (this.isTraced(path, level)) {
            // new

            await Stacktrace.mapFrames(frame)

            // format

            const formattedMessage = message.replace(/{(\d+)}/g, function(match, number) {
                let value = args[+number]

                if (value === undefined) value = "undefined"
                else if (value === null) value = "null"

                return value
            })

            // and write

            this.sink?.trace(new TraceEntry(path, level, formattedMessage, new Date(), frame))
        }
    }

    // private

    private getTraceLevel(path : string) : TraceLevel {
        // check dirty state

        if (this.modifications > 0) {
            this.cachedTraceLevels = {} // restart from scratch
            this.modifications = 0
        } // if

        let level = this.cachedTraceLevels[path]
        if (!level) {
            level = this.traceLevels[path]
            if (!level) {
                const index = path.lastIndexOf(".")
                level = index != -1 ? this.getTraceLevel(path.substring(0, index)) :
                    (path != "" ? this.getTraceLevel("") : TraceLevel.OFF)
            } // if

            // cache

            this.cachedTraceLevels[path] = level
        } // if

        return level
    }

    private setTraceLevel(path : string, level : TraceLevel) : void {
        this.traceLevels[path] = level
        this.modifications++
    }
}

File: common/src/main/kotlin/org/sirius/common/tracer/Tracer.kt
package org.sirius.common.tracer
/*
 * @COPYRIGHT (C) 2023 Andreas Ernst
 *
 * All rights reserved
 */

import java.util.*

/**
 * A Tracer is used to emit trace messages for development purposes.
 * While it shares the logic of a typical logger, it will be turned of in production.
 */
class Tracer(private val trace: Trace, layout : String = "%d %l [%p] %m") {
    // instance data

    private val traceLevels = HashMap<String, TraceLevel>()
    private val cachedTraceLevels = HashMap<String, TraceLevel>()
    private var modifications = 0
    var formatter = TraceFormatter(layout)
    
    init {
        instance = this
    }
    
    // public

    fun isTraced(path: String, level: TraceLevel): Boolean {
        return this.getTraceLevel(path) >= level
    }

    fun trace(path: String, level: TraceLevel, message: String, vararg args: Any) {
        if (this.isTraced(path, level))
            this.trace.trace(TraceEntry(path, level, message.format(*args), Date()), formatter)
    }

    // private

    private fun getTraceLevel(path: String): TraceLevel {
        // check dirty state

        if (this.modifications > 0) {
            this.cachedTraceLevels.clear() // restart from scratch
            this.modifications = 0
        } // if

        var level = this.cachedTraceLevels[path]
        if (level == null) {
            level = this.traceLevels[path]
            if (level == null) {
                val index = path.lastIndexOf(".")
                level =
                    if ( index != -1 )
                        this.getTraceLevel(path.substring(0, index))
                    else (if (path != "") this.getTraceLevel("") else TraceLevel.OFF)
            } // if

            // cache

            this.cachedTraceLevels[path] = level
        } // if

        return level
    }

    fun setTraceLevel(path: String, level: TraceLevel) : Tracer {
        this.traceLevels[path] = level
        this.modifications++

        return this
    }
    
    companion object {
        val ENABLED = System.getProperty("tracing.enabled", "false").lowercase(Locale.getDefault()).equals("true")
        
        private lateinit var instance : Tracer

        // public

        fun isTraced(path: String, level: TraceLevel): Boolean {
            return ENABLED && instance.isTraced(path, level)
        }

        fun trace(path: String, level: TraceLevel, message: String, vararg args: Any) {
            instance.trace(path, level, message, *args)
        }
    }
}